Process command-line arguments as an expression

DurvalMenezes opened the pull request:
Exempli Gratia: 
```
$ ./rpn 
> 1 1 + 
= 2
> q
Bye.
$ ./rpn 1 1 + 
2
$ 
```
Notice the result of evaluating the command-line is printed without any formatting (no "= " prefix, no colors, etc) just so it can be used straight out by other programs/scripts, just as an RPN brother to `expr`.

DurvalMenezes committed:
Process command-line arguments as an expression

marcopaganini commented on main.go:
```diff
@@ -156,20 +156,28 @@ func calc(stack *stackType, cmd string) error {
-	if err := calc(stack, ""); err != nil {
+
+	// treat the cmdline arguments as an expression to be processed
+	expr := ""
```
You can replace these lines with:
```
expr := strings.Join(os.Args[1:])
```
or add the `strings.Join` directly to line 180 below.

DurvalMenezes replied:
> or add the strings.Join directly to line 180 below.

Done, albeit strings.Join() needed an extra argument (the separator): 
```
    if err := calc(stack, strings.Join(os.Args[1:], " ")); err != nil {
        log.Fatal(err)
    }
```

Tested and working, PR updated.

marcopaganini commented on stack.go:
```diff
@@ -51,8 +51,17 @@ func (x *stackType) top() float64 {
 // printTop displays the top of the stack using the base indicated.
-func (x *stackType) printTop(base int) {
-	color.Cyan("= %s", formatNumber(x.top(), base))
+func (x *stackType) printTop(base int, args ...bool) {
```
Is this function working at all? What's the "args ...bool" for? I can't understand what is happening here. There's also a bare `println` that shouldn't work.
 
 If we want to avoid the '=' sign on single commands mode, one option is to pass single to this method and treat it differently. Another way is to not call printTop, instead just print the top of the stack directly with ` `fmt.Println(stack.top())` from outside the function.

DurvalMenezes replied:
> Is this function working at all?

Yep, it's working just great as you can see in the "Exempli Gratia" in my OP :smiley: 

> What's the "args ...bool" for? I can't understand what is happening here.

The idea was to change the code by the minimum possible amount. So, I added the `args ...bool` part to  stack.go/printTop() to     turn it into a variadic function (one accepting a variable number of arguments), and now it can be either called with a 2nd argument to indicate whether to "pretty-print" (with the "> " prefix and colors) the result, or not -- and that way I can call it from main.c/calc(), while all the rest of the code can continue to call it in the exact previous way (which will be handled in the exact same behavior as before, ie, will result in pretty-printing). That way, no changes are needed in any other parts of the code. 

> There's also a bare println that shouldn't work.

But work it does :smile: Again see the "Exempli Gratia" in my OP.

> If we want to avoid the '=' sign on single commands mode, one option is to pass single to this method and treat it differently.

That's exactly what I did from main.go/calc(), the variadic change above is just to avoid changing any other calls elsewhere. But if you prefer it the KISS way, I can remove the variadic and just add a normal parameter, and then change the other calls to add that parameter -- but then I would have to pass `single` somehow on to them too, so not as KISS as you might have planned. 

> Another way is to not call printTop, instead just print the top of the stack directly with  fmt.Println(stack.top())` from outside the function.

I thought about that too, but it would mean 'intruding' into stack.go internals, so not as 'clean' as the other solutions pointed above. 

Please consider my points above and tell me how you prefer it, and I will get it  done :+1:

DurvalMenezes replied:
Done as discussed in our chat, PR updated.

DurvalMenezes committed:
Replaced expr loop with call to strings.Join()

Per upstream suggestion: https://github.com/marcopaganini/rpn/pull/10#discussion_r1770744622

DurvalMenezes committed:
Refactored plain printing.
